common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala

```diff
@@ -144,6 +146,22 @@ trait LoggingSuiteBase
       }
   }
 
+  // An enumeration value that is not defined in `LogKey`, you can define it anywhere
+  private val externalLogKey = LogKey.VALUE
+  private val externalLog = log"${MDC(externalLogKey, "External log message.")}"
+  test("Logging with external LogKey") {
```

Comment:
For this case, the key for `externalLog.context` will be an empty string. Is this expected?

Reply:
In https://github.com/apache/spark/pull/46186/files, I try to relax the type of `MDC.key` to `Enumeration#Value`, which allows us to extend the use of other enumeration types as `MDC.key` to enhance a bit of extensibility.

Reply:
friendly ping @gengliangwang,  what's your take on this? Is `StructuredLogging` a feature that users can utilize? If so, how can we make it easier for users to extend the definition of `LogKey`?

Reply:
@LuciferYang @gengliangwang 
I have updated a version that removes `enumeration`. When using structured logging mechanisms in external Spark related ecosystems, the following format can be used:
```
case object CUSTOME_LOG_KEY extends ILogKey
```
At present, this form is more in line with the traditional Spark style, and we can `constrain` the `class` of MDC parameter `key`

Reply:
Perhaps we also need to consider namespace isolation. Currently, we can define keys with the same name in multiple places, but we cannot distinguish from the `MessageWithContext` whether they are defined internally in Spark or externally.

Reply:
If necessary in the future, I think it would be easier to implement based on the `current changes`, such as
- `ILogKey`

- `InternalLogKey` extends `ILogKey`
- `LogKey` extends `InternalLogKey`

- `ExternalLogKey` extends `ILogKey`
- `ThirdPartyLogKey` extends `ExternalLogKey`

Reply:
We should definitely allow external libraries to extend the log keys. But I don't have a good suggestion for this one now. Either making it as a trait or making the keys as strings are ok.

Reply:
@gengliangwang 
What do you think of the current modification approach?
If we define it as `a string`, it feels like there is `no binding requirement` for `LogKey` at all
Perhaps we need more eyes to review the changes above?

Reply:
Yeah creating a trait seems better. I have some preference in the naming. See my other comments.

Reply:
> If necessary in the future, I think it would be easier to implement based on the `current changes`, such as
> 
> * `ILogKey`
> * `InternalLogKey` extends `ILogKey`
> * `LogKey` extends `InternalLogKey`
> * `ExternalLogKey` extends `ILogKey`
> * `ThirdPartyLogKey` extends `ExternalLogKey`

Yes, we can further expand after this PR is completed. 

Personally, I think we can add a String field to the MDC to identify the namespace. For example, the namespace of Spark's internal MDC can always be `SPARK`, and the third-party Spark ecosystem can customize their own namespace. This can make the key of `MessageWithContext` become similar to `SPARK.mdcKey`, `CUSTOM_NAMESPACE.mdcKey`. 

Perhaps this is not a perfect method, we can further discuss it after this PR is completed.